### ISSUE TITLE
minitest/mock emits "redefining object_id" warning on Ruby head

### ISSUE THREAD

**Issue by mattbrictson** [minitest/minitest#1019]
When I use `minitest/mock` on the latest build of Ruby 3.4.0dev, I see this warning:

> minitest/mock.rb:33: warning: redefining 'object_id' may cause serious problems

This warning can be demonstrated by running minitest's CI workflow on the master branch, specifically the **test ubuntu-latest head** job. Here is an example:

https://github.com/mattbrictson/minitest/actions/runs/11638307079/job/32412919800#step:5:5

_Edit: here's a concise reproduction using just `irb`:_

```
irb(main):001> RUBY_DESCRIPTION
=> "ruby 3.4.0dev (2024-11-01T14:49:50Z master c7708d22c3) +PRISM [arm64-darwin24]"
irb(main):002> require "minitest/mock"
/Users/mbrictson/.rbenv/versions/ruby-dev/lib/ruby/gems/3.4.0+0/gems/minitest-5.25.1/lib/minitest/mock.rb:33: warning: redefining 'object_id' may cause serious problems
=> true
```

**Comment by zenspider** [comment]
Sadly, the best I can come up with is to disable warnings during the offense:

```diff
--- //src/minitest/dev/lib/minitest/mock.rb	2024-07-30 23:54:25.000000000 -0700
+++ /Users/ryan/Work/p4/zss/src/minitest/dev/lib/minitest/mock.rb	2024-07-30 23:54:25.000000000 -0700
@@ -30,6 +30,7 @@
     end
 
     overridden_methods.map(&:to_sym).each do |method_id|
+      old_w, $-w = $-w, nil
       define_method method_id do |*args, **kwargs, &b|
         if @expected_calls.key? method_id then
           if kwargs.empty? then # FIX: drop this after 2.7 dead
@@ -45,6 +46,8 @@
           end
         end
       end
+    ensure
+      $-w = old_w
     end
 
     def initialize delegator = nil # :nodoc:
```

**Comment by zenspider** [comment]
posted a fix
